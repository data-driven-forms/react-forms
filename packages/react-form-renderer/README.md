[![codecov](https://codecov.io/gh/data-driven-forms/react-forms/branch/master/graph/badge.svg)](https://codecov.io/gh/data-driven-forms/react-forms)
[![CircleCI](https://circleci.com/gh/data-driven-forms/react-forms/tree/master.svg?style=svg)](https://circleci.com/gh/data-driven-forms/react-forms/tree/master)
[![npm version](https://badge.fury.io/js/%40data-driven-forms%2Freact-form-renderer.svg)](https://badge.fury.io/js/%40data-driven-forms%2Freact-form-renderer)
[![Tweet](https://img.shields.io/twitter/url/https/github.com/tterb/hyde.svg?style=social)](https://twitter.com/intent/tweet?text=Check%20DataDrivenForms%20React%20library%21%20https%3A%2F%2Fdata-driven-forms.org%2F&hashtags=react,opensource,datadrivenforms)
[![Twitter Follow](https://img.shields.io/twitter/follow/DataDrivenForms.svg?style=social)](https://twitter.com/DataDrivenForms)

[![Data Driven Form logo](https://raw.githubusercontent.com/data-driven-forms/react-forms/master/images/logo.png)](https://data-driven-forms.org/)

Data Driven Forms is a React library used for rendering and managing forms with a lot of provided features based on [React Final Form](https://github.com/final-form/react-final-form).


:question: Why to use Data Driven Forms? :question:
- All forms shared the same functionality!
- Components are controlled in one place!
- You can easily migrate to different sets of components!
- All functionality is provided (see below!)

:tada: Features :tada:
- Easily readable schema, you don't need to know any HTML or JS to be able to write your own form schemas!
- You can use your own components or use one of our provided mappers!
- Form state managing out-of-the-box (including error states!)
- Fully customizable (you can use your own buttons, actions, etc.)!
- Conditional fields!
- Custom field validation with basic set included!
- Datatype validation!
- Cross-field validation!
- Asynchronous validation supported!
- Supporting Wizard forms!
- Supporting [Final Form Field Array](https://github.com/final-form/react-final-form-arrays)!
- ... and a lot more!

:book: For more information please visit the [documentation](https://data-driven-forms.org/). :book:

**Table of Contents**

- [Installation](#installation)
  - [React Form Renderer](#react-form-renderer)
  - [Material-UI Mapper](#material-ui-mapper)
  - [PatternFly 4 Mapper](#patternfly-4-mapper)
  - [BlueprintJS Mapper](#blueprintjs-mapper)
  - [Semantic UI Mapper](#semantic-ui-mapper)
  - [Ant Design Mapper](#ant-design-mapper)
  - [Carbon Mapper](#carbon-mapper)
- [Usage](#usage)
  - [Custom mapper](#custom-mapper)
- [Basic provided components](#basic-provided-components)
- [Documentation](#documentation)
- [Useful links](#useful-links)
- [Development setup](#development-setup)
  - [Tests](#tests)
  - [Commits](#commits)
  - [Changes to documentation](#changes-to-documentation)
- [Contribution](#contribution)
- [LICENSE](#license)

### Installation

Add React Form Renderer

#### [React Form Renderer](https://www.npmjs.com/package/@data-driven-forms/react-form-renderer)

```console
$ npm install @data-driven-forms/react-form-renderer -S
```

```console
$ yarn add @data-driven-forms/react-form-renderer
```

Optionally you can install one of provided mappers:


#### [Material-UI Mapper](https://data-driven-forms.org/mappers/mui-component-mapper)

```console
$ npm install @data-driven-forms/mui-component-mapper -S
```

```console
$ yarn add @data-driven-forms/mui-component-mapper
```

#### [PatternFly 4 Mapper](https://data-driven-forms.org/mappers/pf4-component-mapper)

```console
$ npm install @data-driven-forms/pf4-component-mapper -S
```

```console
$ yarn add @data-driven-forms/pf4-component-mapper
```

#### [BlueprintJS Mapper](https://data-driven-forms.org/mappers/blueprint-component-mapper)

```console
$ npm install @data-driven-forms/blueprint-component-mapper -S
```

```console
$ yarn add @data-driven-forms/blueprint-component-mapper
```

#### [Semantic UI Mapper](https://data-driven-forms.org/mappers/suir-component-mapper)

```console
$ npm install @data-driven-forms/suir-component-mapper -S
```

```console
$ yarn add @data-driven-forms/suir-component-mapper
```

#### [Ant Design Mapper](https://data-driven-forms.org/mappers/ant-component-mapper)

```console
$ npm install @data-driven-forms/ant-component-mapper -S
```

```console
$ yarn add @data-driven-forms/ant-component-mapper
```

#### [Carbon Mapper](https://data-driven-forms.org/mappers/carbon-component-mapper)

```console
$ npm install @data-driven-forms/carbon-component-mapper -S
```

```console
$ yarn add @data-driven-forms/carbon-component-mapper
```

Component libraries in mappers are external dependencies. Make sure to install them in your bundles.

### Usage

In order to Data Driven Forms in your component you need the renderer and a component mapper, which provides component mapper and form template.

```jsx
import React from 'react';
import { FormRenderer, componentTypes } from '@data-driven-forms/react-form-renderer';
import { componentMapper, FormTemplate } from '@data-driven-forms/pf4-component-mapper';

const schema = {
  fields: [{
    component: componentTypes.TEXT_FIELD,
    name: 'name',
    label: 'Your name'
  }]
}

const Form = () => (
  <FormRenderer
    schema={schema}
    componentMapper={componentMapper}
    FormTemplate={FormTemplate}
    onSubmit={console.log}
  />
)
```

#### Custom mapper

You can also use custom mapper.

```jsx
import React from 'react';
import { FormRenderer, componentTypes, useFieldApi } from '@data-driven-forms/react-form-renderer';

const TextField = props => {
  const {label, input, meta, ...rest} = useFieldApi(props)
  return (
    <div>
      <label htmlForm={input.name}>{label}</label>
      <input {...input} {...rest} id={input.name}/>
      {meta.error && <p>{meta.error}</p>}
    </div>
  )
}

const componentMapper = {
  [componentTypes.TEXT_FIELD]: TextField,
  'custom-type': TextField
}

const schema = {
  fields: [{
    component: componentTypes.TEXT_FIELD,
    name: 'name',
    label: 'Your name'
    type: 'search',
  }]
}
```

For more information, please check [this page](https://data-driven-forms.org/renderer/component-mapping).

Mappers can be also generated by using `yarn generate-template` [command](https://data-driven-forms.org/renderer/development-setup#generatingamappertemplate).

### Basic provided components

Data Driven Forms supports all kinds of component, basic set is consisted of:

- [Text input](https://data-driven-forms.org/mappers/text-field?mapper=mui)
- [Text area](https://data-driven-forms.org/mappers/textarea?mapper=mui)
- [Checkbox](https://data-driven-forms.org/mappers/checkbox?mapper=mui) ([Multiple checkboxes](https://data-driven-forms.org/mappers/checkbox-multiple?mapper=mui))
- [Select (dropdown)](https://data-driven-forms.org/mappers/select?mapper=mui)
- [Dual list select](https://data-driven-forms.org/mappers/dual-list-select?mapper=mui)
- [Field array](https://data-driven-forms.org/mappers/field-array?mapper=mui)
- [Switch](https://data-driven-forms.org/mappers/switch?mapper=mui)
- [Radio buttons](https://data-driven-forms.org/mappers/radio?mapper=mui)
- [Date picker](https://data-driven-forms.org/mappers/date-picker?mapper=mui)
- [Time picker](https://data-driven-forms.org/mappers/time-picker?mapper=mui)
- [Tabs](https://data-driven-forms.org/mappers/tabs?mapper=mui)
- [Slider](https://data-driven-forms.org/mappers/slider?mapper=mui)
- [Sub-form](https://data-driven-forms.org/mappers/sub-form?mapper=mui)
- [Plain text](https://data-driven-forms.org/mappers/plain-text?mapper=mui)
- [Wizard](https://data-driven-forms.org/mappers/wizard?mapper=mui)

Any other components can be added to mapper and renderer with the form renderer. Existing components can be also overriden.

### Documentation

Please use our [documentation site](https://data-driven-forms.org/). In case of any problem, you can access documentation files directly in GitHub.

### Useful links

- [Data Driven Forms documentation](https://data-driven-forms.org/)
- [PatternFly 4 Design documentation](https://www.patternfly.org/v4/)
- [Material-UI documentation](https://mui.com/)
- [Ant Design documentation](https://ant.design/)
- [Semantic UI React](https://react.semantic-ui.com/)
- [BlueprintJS](https://blueprintjs.com/)
- [IBM Carbon](https://www.carbondesignsystem.com/)
- NPM
  - [React Form Renderer](https://www.npmjs.com/package/@data-driven-forms/react-form-renderer)
  - [PatternFly 4 Mapper](https://www.npmjs.com/package/@data-driven-forms/pf4-component-mapper)
  - [MaterialUI Mapper](https://www.npmjs.com/package/@data-driven-forms/mui-component-mapper)
  - [Ant Design Mapper](https://www.npmjs.com/package/@data-driven-forms/ant-component-mapper)
  - [Semantic UI Mapper](https://www.npmjs.com/package/@data-driven-forms/suir-component-mapper)
  - [BlueprintJS Mapper](https://www.npmjs.com/package/@data-driven-forms/blueprint-component-mapper)
  - [Carbon Mapper](https://www.npmjs.com/package/@data-driven-forms/carbon-component-mapper)

### Development setup

Data Driven Forms is a monorepo that uses [Lerna](https://github.com/lerna/lerna) and [yarn workspaces](https://classic.yarnpkg.com/blog/2017/08/02/introducing-workspaces/), so you can use all its commands as well.

1. Install

```console
yarn install
```

2. Build

```console
yarn build
```

3. Run a package

Each package has a small playground `package/demo`, where you can test your changes.

```console
cd packages/pf3-component-mapper
yarn start
```

4. How to clean?

```console
yarn lerna clean # will delete all node_modules
```

All packages are linked together by default, so if you run a `yarn build` in a package, all other packages are updated to the latest version of that package.

#### Tests

Tests can be ran from the core folder or from specific packages.

```console
yarn test

yarn test packages/pf3-component-mapper
```

#### Commits

Data Driven Forms uses [Semantic Release](https://github.com/semantic-release/commit-analyzer)

Format:

```
[type]([package]): message

fix(pf3): title accepts node
```

Types:
- `feat`: a new feature, will trigger new `_.X._` release
- `fix`: a fix, will trigger new `_._.X` release

Packages:
- Please describe which package is being changed `pf3`, `renderer`, ...

Please, do not use Semantic Release, if you update only the demo.

All packages are releasing together and they share the version number.

#### Changes to documentation

If your changes influence API or add new features, you should describe these new options in the `react-renderer-demo` repository. Thanks!

### Contribution

We welcome any community contribution. Don't be afraid to report bug or to create issues and pull-requests! :trophy:

### LICENSE

Apache License 2.0
