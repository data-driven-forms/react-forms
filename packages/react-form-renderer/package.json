{
  "name": "@data-driven-forms/react-form-renderer",
  "version": "3.2.0",
  "description": "React form renderer for data-driven-forms.",
  "main": "index.js",
  "module": "esm/index.js",
  "license": "Apache-2.0",
  "scripts": {
    "start": "webpack-dev-server --env dev --config ./config/webpack.config.js --open --hot",
    "build": "yarn build:cjs && yarn build:esm && yarn build:typings && yarn build:packages",
    "build:cjs": "BABEL_ENV=cjs babel src --out-dir ./ --ignore \"src/tests/*\"",
    "build:esm": "BABEL_ENV=esm babel src --out-dir ./esm --ignore \"src/tests/*\"",
    "build:typings": "node ../../scripts/generate-typings.js",
    "build:packages": "node ../../scripts/generate-packages.js",
    "release": "semantic-release"
  },
  "repository": "git@github.com:data-driven-forms/react-forms.git",
  "keywords": [
    "react",
    "forms",
    "javascript"
  ],
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/plugin-proposal-class-properties": "^7.1.0",
    "@babel/plugin-syntax-dynamic-import": "^7.0.0",
    "@babel/preset-env": "^7.1.6",
    "@babel/preset-react": "^7.0.0",
    "@semantic-release/git": "^7.0.5",
    "@semantic-release/npm": "^5.1.1",
    "babel-core": "^7.0.0-bridge.0",
    "babel-eslint": "9.x",
    "babel-jest": "^23.6.0",
    "babel-loader": "^8.0.4",
    "babel-plugin-lodash": "^3.3.4",
    "css-loader": "^1.0.1",
    "enzyme": "^3.7.0",
    "enzyme-adapter-react-16": "^1.7.0",
    "enzyme-to-json": "^3.3.4",
    "glob": "^7.1.6",
    "html-webpack-plugin": "^3.2.0",
    "mini-css-extract-plugin": "^0.4.4",
    "node-sass": "^4.10.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "regenerator-runtime": "^0.12.1",
    "resolve-url-loader": "^3.0.0",
    "rollup": "^1.23.1",
    "rollup-plugin-babel": "^4.3.3",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-globals": "^1.4.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-sourcemaps": "^0.5.0",
    "rollup-plugin-terser": "^5.1.2",
    "sass-loader": "^7.1.0",
    "semantic-release": "^17.2.3",
    "style-loader": "^0.23.1",
    "url-loader": "^1.1.2",
    "webpack": "^4.25.1",
    "webpack-cli": "^3.1.2",
    "webpack-dev-server": "^3.1.10",
    "webpack-merge": "^4.1.4"
  },
  "dependencies": {
    "lodash": "^4.17.15",
    "prop-types": "^15.7.2",
    "react-final-form": "^6.5.0",
    "react-final-form-arrays": "^3.1.1"
  },
  "peerDependencies": {
    "react": ">=16.13.1",
    "react-dom": ">=16.13.1"
  },
  "postpublish": "export RELEASE_DEMO=true"
}
