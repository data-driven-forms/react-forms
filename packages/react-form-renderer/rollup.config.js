import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import replace from 'rollup-plugin-replace';
import nodeGlobals from 'rollup-plugin-node-globals';
import { terser } from 'rollup-plugin-terser';
import typescript from '@rollup/plugin-typescript';
import glob from 'glob';
import path from 'path';
import sourcemaps from 'rollup-plugin-sourcemaps';

const outputPaths = [...glob.sync(path.resolve(__dirname, './src/files/!(*.d.ts)')), ...glob.sync(path.resolve(__dirname, './src/files/*.js'))];

const globals = {
  react: 'React',
  'react-dom': 'ReactDOM'
};

const babelOptions = {
  exclude: /node_modules/,
  runtimeHelpers: true,
  configFile: '../../babel.config.js'
};

const commonjsOptions = {
  ignoreGlobal: true,
  include: /node_modules/,
  namedExports: {
    './src/form-renderer/helpers': ['composeValidators']
  }
};

function onwarn(warning) {
  throw Error(warning.message);
}

const plugins = [
  nodeResolve(),
  babel(babelOptions),
  commonjs(commonjsOptions),
  nodeGlobals(),
  replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
  terser({
    keep_classnames: true,
    keep_fnames: true
  }),
  typescript({
    allowSyntheticDefaultImports: true
  }),
  sourcemaps()
];

export default [
  ...['cjs', 'esm'].map((env) => ({
    onwarn,
    input: ['./src/index.js', ...outputPaths],
    output: {
      dir: `./dist/${env}`,
      format: env,
      name: '@data-driven-forms/react-form-renderer',
      exports: 'named',
      globals,
      sourcemap: true
    },
    external: Object.keys(globals),
    plugins
  })),
  {
    onwarn,
    input: './src/index.js',
    output: {
      file: `./dist/umd/index.js`,
      format: 'umd',
      name: '@data-driven-forms/react-form-renderer',
      exports: 'named',
      globals,
      sourcemap: true
    },
    external: Object.keys(globals),
    plugins
  }
];
