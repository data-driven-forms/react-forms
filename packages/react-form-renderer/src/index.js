export { default as componentTypes } from './component-types';
export { default as composeValidators } from './compose-validators';
export { default as condition } from './condition';
export { default as dataTypes } from './data-types';
export { default as defaultSchemaValidator } from './default-schema-validator';
export { default as FieldArray } from './field-array';
export { default as FieldProvider } from './field-provider';
export { default as FormRenderer } from './form-renderer';
export { default as FormSpy } from './form-spy';
export { default as Form } from './form';
export { default as FormError } from './form-error';
export { default as parseCondition } from './parse-condition';
export { default as RendererContext } from './renderer-context';
export { default as schemaErrors } from './schema-errors';
export { default as useFieldApi } from './use-field-api';
export { default as useFormApi } from './use-form-api';
export { default as validatorMapper } from './validator-mapper';
export { default as validatorTypes } from './validator-types';
export { default as validators } from './validators';
export { default as wizardContext } from './wizard-context';
