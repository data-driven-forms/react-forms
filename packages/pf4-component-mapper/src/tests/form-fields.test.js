import React from 'react';
import { act } from 'react-dom/test-utils';
import toJson from 'enzyme-to-json';
import { FormGroup, Radio as PF4Radio } from '@patternfly/react-core';
import { mount } from 'enzyme';
import Checkbox from '../checkbox';
import Switch from '../switch';

import RenderWithProvider from '../../../../__mocks__/with-provider';
import { FormRenderer, componentTypes } from '@data-driven-forms/react-form-renderer';
import FormTemplate from '../form-template';
import componentMapper from '../component-mapper';
import { validatorTypes } from '@data-driven-forms/react-form-renderer';
import MultipleChoiceListCommon from '@data-driven-forms/common/multiple-choice-list';

describe('FormFields', () => {
  const props = {
    name: 'Name of the field',
    id: 'someIdKey'
  };

  it('should render with onText/OffText Switch correctly', () => {
    const wrapper = mount(
      <RenderWithProvider>
        <Switch {...props} onText="I am on" offText="Turned off" />
      </RenderWithProvider>
    );

    expect(
      wrapper
        .find('.pf-m-on')
        .text()
        .includes('I am on')
    ).toEqual(true);
    expect(
      wrapper
        .find('.pf-m-on')
        .text()
        .includes('Turned off')
    ).toEqual(false);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('formFields generated tests', () => {
    const RendererWrapper = ({ schema = { fields: [] }, ...props }) => (
      <FormRenderer
        onSubmit={jest.fn()}
        FormTemplate={(props) => <FormTemplate {...props} />}
        schema={schema}
        componentMapper={componentMapper}
        subscription={{ submitFailed: true }}
        {...props}
      />
    );
    let field;
    let schema;

    const errorText = 'Required';
    const helperText = 'I am helper text';
    const description = 'This is description';
    const options = [
      { label: 'Cat', value: 'cats' },
      { label: 'Dog', value: 'dogs' },
      { label: 'Hamster', value: 'hamsters' }
    ];

    const componentsWithOptions = [componentTypes.RADIO, componentTypes.SELECT];

    beforeEach(() => {
      field = {};
      schema = [];
    });

    describe('components', () => {
      [
        componentTypes.TEXT_FIELD,
        componentTypes.TEXTAREA,
        componentTypes.RADIO,
        componentTypes.CHECKBOX,
        componentTypes.DATE_PICKER,
        componentTypes.TIME_PICKER,
        componentTypes.SWITCH,
        componentTypes.SELECT,
        componentTypes.SLIDER
      ].forEach((component) => {
        describe(`Component type: ${component}`, () => {
          beforeEach(() => {
            field.component = component;
            field.name = 'field-name';
            field.label = 'Some label';

            if (componentsWithOptions.includes(component)) {
              field = {
                ...field,
                options
              };
            }

            schema = { fields: [field] };
          });

          it('renders correctly', () => {
            const wrapper = mount(<RendererWrapper schema={schema} />);

            if (component === componentTypes.RADIO) {
              expect(wrapper.find(PF4Radio)).toHaveLength(options.length);
            } else if (component === componentTypes.SWITCH) {
              expect(
                wrapper
                  .find('.pf-c-switch__label')
                  .first()
                  .text()
              ).toEqual(field.label);
            } else {
              expect(wrapper.find(componentMapper[component])).toHaveLength(1);
              expect(wrapper.find('label').text()).toEqual(field.label);
            }

            expect(wrapper.find(FormGroup)).toHaveLength(1);
            expect(wrapper.find('.pf-m-error')).toHaveLength(0);
            expect(wrapper.find('.pf-c-form__helper-text')).toHaveLength(0);
            expect(wrapper.find('.pf-c-form__label-required')).toHaveLength(0);
          });

          it('renders with error', () => {
            const errorField = {
              ...field,
              validate: [{ type: validatorTypes.REQUIRED }]
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [errorField] }} />);
            wrapper.find('form').simulate('submit');
            expect(
              wrapper
                .find('.pf-m-error')
                .last()
                .text()
            ).toEqual(errorText);
          });

          it('renders with error and validateOnMount', async () => {
            const errorField = {
              ...field,
              validate: [{ type: validatorTypes.REQUIRED }],
              validateOnMount: true
            };
            let wrapper;
            await act(async () => {
              wrapper = mount(<RendererWrapper schema={{ fields: [errorField] }} />);
            });
            wrapper.update();
            expect(
              wrapper
                .find('.pf-m-error')
                .last()
                .text()
            ).toEqual(errorText);
          });

          it('renders with helperText', () => {
            const helpertextField = {
              ...field,
              helperText
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [helpertextField] }} />);

            expect(
              wrapper
                .find('.pf-c-form__helper-text')
                .last()
                .text()
            ).toEqual(helperText);
          });

          it('renders with description', () => {
            const descriptionField = {
              ...field,
              description
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [descriptionField] }} />);

            expect(
              wrapper
                .find('small')
                .last()
                .text()
            ).toEqual(description);
          });

          it('renders with description and helperText', () => {
            const descriptionField = {
              ...field,
              description,
              helperText
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [descriptionField] }} />);

            expect(
              wrapper
                .find('.pf-c-form__helper-text')
                .last()
                .text()
            ).toEqual(helperText);
            expect(
              wrapper
                .find('small')
                .last()
                .text()
            ).toEqual(description);
          });

          if (![componentTypes.SELECT, componentTypes.SLIDER, componentTypes.RADIO].includes(component)) {
            it('renders with warning and helperText', async () => {
              const errorFields = {
                ...field,
                helperText,
                id: 'warning-field',
                validate: [() => ({ type: 'warning', error: errorText })]
              };

              let wrapper;

              await act(async () => {
                wrapper = mount(<RendererWrapper schema={{ fields: [errorFields] }} />);
              });
              wrapper.update();

              await act(async () => {
                wrapper
                  .find('#warning-field')
                  .last()
                  .simulate('focus');
              });
              wrapper.update();

              await act(async () => {
                wrapper
                  .find('#warning-field')
                  .last()
                  .simulate('blur');
              });
              wrapper.update();

              expect(
                wrapper
                  .find('.pf-m-warning')
                  .last()
                  .text()
              ).toEqual(errorText);
            });
          }

          it('renders isRequired', () => {
            const requiredField = {
              ...field,
              isRequired: true
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [requiredField] }} />);

            if (component === componentTypes.SWITCH) {
              expect(wrapper.find('.pf-c-form__label-required')).toHaveLength(2);
            } else {
              expect(wrapper.find('.pf-c-form__label-required')).toHaveLength(1);
            }
          });

          it('renders isDisabled', () => {
            if (component === componentTypes.SLIDER) {
              return;
            }

            const disabledField = {
              ...field,
              isDisabled: true
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [disabledField] }} />);

            if (component === componentTypes.TEXTAREA) {
              expect(
                wrapper
                  .find('textarea')
                  .first()
                  .props().disabled
              ).toEqual(true);
            } else if (component === componentTypes.SELECT) {
              expect(wrapper.find('div.pf-c-select__toggle').prop('disabled')).toEqual(true);
            } else {
              expect(
                wrapper
                  .find('input')
                  .first()
                  .props().disabled
              ).toEqual(true);
            }
          });

          it('renders isReadOnly', () => {
            if (component === componentTypes.SELECT || component === componentTypes.SLIDER) {
              return;
            }

            const disabledField = {
              ...field,
              isReadOnly: true
            };
            const wrapper = mount(<RendererWrapper schema={{ fields: [disabledField] }} />);

            if (component === componentTypes.TEXTAREA) {
              expect(
                wrapper
                  .find('textarea')
                  .first()
                  .props().disabled
              ).toEqual(true);
            } else if ([componentTypes.CHECKBOX, componentTypes.RADIO, componentTypes.SWITCH, componentTypes.SLIDER].includes(component)) {
              expect(
                wrapper
                  .find('input')
                  .first()
                  .props().disabled
              ).toEqual(true);
            } else {
              expect(
                wrapper
                  .find('input')
                  .first()
                  .props().readOnly
              ).toEqual(true);
            }
          });

          it('renders with submit error', async () => {
            const wrapper = mount(<RendererWrapper schema={schema} onSubmit={() => ({ [field.name]: errorText })} />);

            await act(async () => {
              wrapper.find('form').simulate('submit');
            });
            wrapper.update();

            expect(
              wrapper
                .find('.pf-m-error')
                .last()
                .text()
            ).toEqual(errorText);
          });
        });
      });
    });

    describe('MultipleCheckbox', () => {
      let initialProps;
      beforeEach(() => {
        initialProps = {
          ...initialProps,
          name: 'checkbox',
          options: [
            { label: 'Cat', value: 'cats' },
            { label: 'Dog', value: 'dogs' },
            { label: 'Hamster', value: 'hamsters' }
          ]
        };
      });

      it('renders correctly', () => {
        const wrapper = mount(
          <RenderWithProvider>
            <Checkbox {...initialProps} />
          </RenderWithProvider>
        );

        expect(wrapper.find(MultipleChoiceListCommon)).toHaveLength(1);
        expect(wrapper.find('.pf-m-error')).toHaveLength(0);
        expect(wrapper.find('.pf-c-form__helper-text')).toHaveLength(0);
        expect(wrapper.find('.pf-c-form__label-required')).toHaveLength(0);
      });

      it('renders with error', () => {
        const schema = {
          fields: [
            {
              component: componentTypes.CHECKBOX,
              name: 'foo',
              validate: [{ type: validatorTypes.REQUIRED }]
            }
          ]
        };
        const wrapper = mount(<RendererWrapper schema={schema} />);
        wrapper.find('form').simulate('submit');

        expect(
          wrapper
            .find('.pf-m-error')
            .last()
            .text()
        ).toEqual(errorText);
      });

      it('renders with helperText', () => {
        const schema = {
          fields: [
            {
              component: componentTypes.CHECKBOX,
              name: 'foo',
              helperText
            }
          ]
        };
        const wrapper = mount(<RendererWrapper schema={schema} />);

        expect(
          wrapper
            .find('.pf-c-form__helper-text')
            .last()
            .text()
        ).toEqual(helperText);
      });

      it('renders with description', () => {
        const schema = {
          fields: [
            {
              component: componentTypes.CHECKBOX,
              name: 'foo',
              description
            }
          ]
        };
        const wrapper = mount(<RendererWrapper schema={schema} />);

        expect(
          wrapper
            .find('small')
            .last()
            .text()
        ).toEqual(description);
      });

      it('renders with description and helperText', () => {
        const schema = {
          fields: [
            {
              component: componentTypes.CHECKBOX,
              name: 'foo',
              helperText,
              description
            }
          ]
        };
        const wrapper = mount(<RendererWrapper schema={schema} />);

        expect(
          wrapper
            .find('.pf-c-form__helper-text')
            .last()
            .text()
        ).toEqual(helperText);
        expect(
          wrapper
            .find('small')
            .last()
            .text()
        ).toEqual(description);
      });

      it('renders with error and helperText', () => {
        const schema = {
          fields: [
            {
              component: componentTypes.CHECKBOX,
              name: 'foo',
              helperText,
              validate: [{ type: validatorTypes.REQUIRED }]
            }
          ]
        };
        const wrapper = mount(<RendererWrapper schema={schema} />);
        wrapper.find('form').simulate('submit');

        expect(
          wrapper
            .find('.pf-m-error')
            .last()
            .text()
        ).toEqual(errorText);
      });
    });
  });
});
