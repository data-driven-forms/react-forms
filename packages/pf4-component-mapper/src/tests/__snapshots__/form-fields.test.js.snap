// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFields should render with onText/OffText Switch correctly 1`] = `
<RenderWithProvider>
  <Component
    onSubmit={[Function]}
  >
    <Switch
      id="someIdKey"
      name="Name of the field"
      offText="Turned off"
      onText="I am on"
    >
      <FormGroup
        hideLabel={true}
        id="someIdKey"
        meta={
          Object {
            "active": false,
            "data": Object {},
            "dirty": false,
            "dirtySinceLastSubmit": false,
            "error": undefined,
            "initial": undefined,
            "invalid": false,
            "modified": false,
            "modifiedSinceLastSubmit": false,
            "pristine": true,
            "submitError": undefined,
            "submitFailed": false,
            "submitSucceeded": false,
            "submitting": false,
            "touched": false,
            "valid": true,
            "validating": false,
            "visited": false,
            "warning": undefined,
          }
        }
      >
        <FormGroup
          fieldId="someIdKey"
          label={false}
          validated="default"
        >
          <div
            className="pf-c-form__group"
          >
            <div
              className="pf-c-form__group-control"
            >
              <Switch
                aria-label=""
                checked={false}
                id="someIdKey"
                isChecked={true}
                isDisabled={false}
                label="I am on"
                labelOff="Turned off"
                name="Name of the field"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox"
                value=""
              >
                <label
                  className="pf-c-switch"
                  data-ouia-component-id="OUIA-Generated-Switch-1"
                  data-ouia-component-type="PF4/Switch"
                  data-ouia-safe={true}
                  htmlFor="someIdKey"
                >
                  <input
                    aria-label=""
                    aria-labelledby="someIdKey-on"
                    checked={false}
                    className="pf-c-switch__input"
                    disabled={false}
                    id="someIdKey"
                    name="Name of the field"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="checkbox"
                    value=""
                  />
                  <span
                    className="pf-c-switch__toggle"
                  />
                  <span
                    aria-hidden="true"
                    className="pf-c-switch__label pf-m-on"
                    id="someIdKey-on"
                  >
                    I am on
                  </span>
                  <span
                    aria-hidden="true"
                    className="pf-c-switch__label pf-m-off"
                    id="someIdKey-off"
                  >
                    Turned off
                  </span>
                </label>
              </Switch>
            </div>
          </div>
        </FormGroup>
      </FormGroup>
    </Switch>
  </Component>
</RenderWithProvider>
`;
